File: wilsonChan/assignments/slidingwindow/lc209/lc209.py
# Problem Statement #

# Given an array of positive numbers and a positive number ‘S,’ find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’. Return 0 if no such subarray exists.

import math


def smallest_subarray_with_given_sum(s, arr):

  #Initialized variables
  window_start = 0        #left/beginning of the window
  window_end = 0          #right/end of the window
  current_sum = 0
  min_length = len(arr) + 1

  #Window continues moving to the right as long it's less than whole array
  while window_end < len(arr):
    current_sum += arr[window_end]
    
    while current_sum >= s:
      min_length = min(min_length, window_end - (window_start + 1))
      current_sum -= arr[window_start]
      window_start += 1

    window_end += 1

  return current_sum if current_sum <= len(arr) else 0

def main():
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(7, [2, 1, 5, 2, 3, 2])))
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(7, [2, 1, 5, 2, 8])))
  print("Smallest subarray length: " + str(smallest_subarray_with_given_sum(8, [3, 4, 1, 1, 6])))


main()
















  # Solution
  # -----
  # window_sum = 0
  # min_length = math.inf
  # window_start = 0

  # for window_end in range(0, len(arr)):
  #   window_sum += arr[window_end]  # add the next element
  #   # shrink the window as small as possible until the 'window_sum' is smaller than 's'
  #   while window_sum >= s:
  #     min_length = min(min_length, window_end - window_start + 1)
  #     window_sum -= arr[window_start]
  #     window_start += 1
  # if min_length == math.inf:
  #   return 0
  # return min_length

File: janMarkusMilan/assignments/slidingwindow/lc209/lc209.js
// Problem Statement #

// Given an array of positive numbers and a positive number ‘S,’ find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’. Return 0 if no such subarray exists.

/*
Edge cases: return 0
Variables: min, sum

Form a window (loop):
    increment sum until >= target
    while sum >= target:
        keep track of current minimum
        subtract leftmost
        left++;
    
    if sum does not equal target:
        return 0

return the min
*/

function smallest_subarray_with_given_sum(s, arr) {
  // TODO: Write code here
  let sum = 0;
  let min = arr.length + 1;

  // Form a 'window' using two pointers over a portion of the data
  let leftPtr = 0;
  for (let rightPtr = 0; rightPtr < arr.length; rightPtr++) {
    // Add current element to the sum (keep track of sum of values in window)
    sum += arr[rightPtr];

    // Once the sum of the values in the window reaches the target, slide the window over by 1
    while (sum >= s) {
      // Keep track of the min of each window slide iteration
      min = Math.min(min, rightPtr - leftPtr + 1);

      // Subtract the left-most value in the window from the sum (since we slide the window forward)
      sum -= arr[leftPtr];

      // Increment the left pointer by 1 (update the beginning position of the window)
      leftPtr++;
    }
  }

  // If the min remains unchanged (the sum of the values in the array don't reach the target), return 0
  if (min === arr.length + 1) {
    return 0;
  }

  return min;
}

console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 3, 2]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    7,
    [2, 1, 5, 2, 8]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    8,
    [3, 4, 1, 1, 6]
  )}`
);
console.log(
  `Smallest subarray length: ${smallest_subarray_with_given_sum(
    100,
    [3, 4, 1, 1, 6]
  )}`
);

// Solution
// -----
// let windowSum = 0,
// minLength = Infinity,
// windowStart = 0;

// for (windowEnd = 0; windowEnd < arr.length; windowEnd++) {
// windowSum += arr[windowEnd]; // add the next element
// // shrink the window as small as possible until the 'window_sum' is smaller than 's'
// while (windowSum >= s) {
//   minLength = Math.min(minLength, windowEnd - windowStart + 1);
//   windowSum -= arr[windowStart];
//   windowStart += 1;
// }
// }

// if (minLength === Infinity) {
// return 0;
// }
// return minLength;
